sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/ErrorHandler.java

```diff
@@ -0,0 +1,187 @@
+ * }</pre>
+ *     Usage with multiple DLQ stages
+ *     <pre>{@code
+ * PCollection<?> records = ...;
+ * try (ErrorHandler<E,T> errorHandler = pipeline.registerErrorHandler(SomeSink.write())) {
+ *  PCollection<?> results = records.apply(SomeIO.write().withDeadLetterQueue(errorHandler))
+ *                        .apply(OtherTransform.builder().withDeadLetterQueue(errorHandler));
+ * }
+ * results.apply(SomeOtherTransform);
+ * }</pre>
+ */
+public interface ErrorHandler<ErrorT, OutputT extends POutput> extends AutoCloseable {
+
+  void addErrorCollection(PCollection<ErrorT> errorCollection);
+
+  boolean isClosed();
+
+  OutputT getOutput();
+
+  class PTransformErrorHandler<ErrorT, OutputT extends POutput>
+      implements ErrorHandler<ErrorT, OutputT> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PTransformErrorHandler.class);
+    private final PTransform<PCollection<ErrorT>, OutputT> sinkTransform;
+
+    private final List<PCollection<ErrorT>> errorCollections = new ArrayList<>();
+
+    private @Nullable OutputT sinkOutput = null;
+
+    private boolean closed = false;
+
+    /**
+     * Constructs a new ErrorHandler, but should not be called directly. Instead, call
+     * pipeline.registerErrorHandler to ensure safe pipeline construction
+     */
+    @Internal
+    public PTransformErrorHandler(PTransform<PCollection<ErrorT>, OutputT> sinkTransform) {
+      this.sinkTransform = sinkTransform;
+    }
+
+    @Override
+    public void addErrorCollection(PCollection<ErrorT> errorCollection) {
+      errorCollections.add(errorCollection);
+    }
+
+    @Override
+    public boolean isClosed() {
+      return closed;
+    }
+
+    @Override
+    public OutputT getOutput() {
+      if (!this.isClosed()) {
+        throw new IllegalStateException(
+            "ErrorHandler must be finalized before the output can be returned");
+      }
+      // make the static analysis checker happy
+      Preconditions.checkArgumentNotNull(sinkOutput);
+      return sinkOutput;
+    }
+
+    @Override
+    public void close() {
+      closed = true;
+      if (errorCollections.isEmpty()) {
+        LOG.warn("Empty list of error pcollections passed to ErrorHandler.");
+        return;
+      }
+      LOG.debug(
+          "{} error collections are being sent to {}",
+          errorCollections.size(),
+          sinkTransform.getName());
+      sinkOutput =
+          PCollectionList.of(errorCollections)
+              .apply(Flatten.pCollections())
+              .apply(new WriteErrorMetrics<ErrorT>(sinkTransform))
+              .apply(
+                  sinkTransform.addAnnotation(
+                      "FeatureMetric", "ErrorHandler".getBytes(StandardCharsets.UTF_8)));
+    }
+
+    public static class WriteErrorMetrics<ErrorT>
+        extends PTransform<PCollection<ErrorT>, PCollection<ErrorT>> {
+
+      private final Counter errorCounter;
+
+      public WriteErrorMetrics(PTransform<?, ?> sinkTransform) {
```

Comment:
Why is sinkTransform being passed here? (Creating a whole Transform for this seems overkill--perhaps apply the ParDo directly in the close method above. Alternatively, perhaps the Flatten + WriteErrorMetrics + ApplySink should be nested in a single transform.)

Reply:
we are just using the sink transform name. I'll clean it up